lib/rust/ensogl/component/slider/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+    fn init(self) -> Self {
+        let network = self.frp.network();
+        let input = &self.frp.input;
+        let output = &self.frp.private.output;
+        let model = &self.model;
+
+        let scene = &self.app.display.default_scene;
+        let mouse = &scene.mouse.frp;
+        let keyboard = &scene.keyboard.frp;
+
+        let track_pos = Animation::new_non_init(network);
+
+        let background_color = color::Animation::new(network);
+        let track_color = color::Animation::new(network);
+        let value_color = color::Animation::new(network);
+        let label_color = color::Animation::new(network);
+
+        frp::extend! { network
+
+            // User input
+
+            background_click        <- model.background.events.mouse_down_primary.constant(());
+            background_release      <- model.background.events.mouse_release_primary.constant(());
+            background_drag         <- bool(&background_release, &background_click);
+
+            track_click             <- model.track.events.mouse_down_primary.constant(());
+            track_release           <- model.track.events.mouse_release_primary.constant(());
+            track_drag              <- bool(&track_release, &track_click);
+
+            component_click         <- any2(&background_click, &track_click).gate(&input.set_slider_enabled);
+            component_drag          <- any2(&background_drag, &track_drag).gate(&input.set_slider_enabled);
+            component_release       <- any2(&background_release, &track_release);
+
+            component_ctrl_click    <- component_click.gate(&keyboard.is_control_down);
+
+            drag_pos_start          <- mouse.position.sample(&component_click);
+            drag_pos_end            <- mouse.position.gate(&component_drag);
+            drag_pos_end            <- any2(&drag_pos_end, &drag_pos_start);
+            drag_delta              <- all2(&drag_pos_end, &drag_pos_start).map(|(end, start)| end - start);
+
+            mouse_y_local           <- mouse.position.map(
+                f!([scene, model] (pos) scene.screen_to_object_space(&model.background, *pos).y )
+            ).gate(&component_drag);
+
+
+
+            // Componenet size
+
+            output.width            <+ input.set_width;
+            output.height           <+ input.set_height;
+
+
+            // precision calculation
+
+            precision_adjust_margin <- all2(&input.set_height, &input.set_precision_step_margin).map(
+                |(height, margin)| height / 2.0 + margin
+            );
+            precision_adjusted      <- all4(
+                &input.set_precision,
+                &mouse_y_local,
+                &precision_adjust_margin,
+                &input.set_precision_step_size,
+            ).map(
+                |(default, offset, margin, step_size)| {
+                    let steps = ((offset.abs() - margin) / step_size).max(0.0).ceil() * offset.signum();
```

Comment:
This is a super complex expression. Please refactor it to well named variables, explaining what these computations mean.